### ISSUE TITLE
Discriminated union not correctly narrowed by intersection

### ISSUE THREAD

**Issue by jtbandes** [microsoft/TypeScript#36323]
I am filing this issue to re-open the discussion from #13300 and #18210 which are now locked. While the former was closed as a duplicate of the latter, which itself was closed in #18438, this change did not actually fix the original issue in #13300. Here is an example:

```ts
type A = { key: "one", value: number } | { key: "two", value: string };

type A_one = A & { key: "one" };

function foo(a1: A_one): number {
    return a1.value;  // error
}
```

`A_one` should be typed as `{ key: "one", value: number }`, because there are no other possibilities, but is actually typed as `({ key: "one", value: number } & { key: "one" }) | ({ key: "two", value: string } & { key: "one" })`.

----

Motivation: https://stackoverflow.com/q/59835356/23649; code I was trying to write to solve this problem:

```ts
function foo<K extends UnionA["key"]>(val: UnionA & { key: K }): (UnionB & { key: K }) | undefined {
    const k: K = val.key;  // not strictly necessary, but just to help reduce complexity below
    for (let i = 0; i < b.length; i++) {
        if (b[i].key === k) {
            return b[i];
        }
    }
    return undefined;
}
```

Error message on the `return` line:
```
Type 'UnionB' is not assignable to type '({ key: "one"; value: boolean; } & { key: K; }) | ({ key: "two"; value: string; } & { key: K; }) | undefined'.
  Type '{ key: "one"; value: boolean; }' is not assignable to type '({ key: "one"; value: boolean; } & { key: K; }) | ({ key: "two"; value: string; } & { key: K; }) | undefined'.
    Type '{ key: "one"; value: boolean; }' is not assignable to type '{ key: "one"; value: boolean; } & { key: K; }'.
      Type '{ key: "one"; value: boolean; }' is not assignable to type '{ key: K; }'.
        Types of property 'key' are incompatible.
          Type '"one"' is not assignable to type 'K'.
            '"one"' is assignable to the constraint of type 'K', but 'K' could be instantiated with a different subtype of constraint '"one" | "two"'.(2322)
```

**Comment by jack-williams** [comment]
@jtbandes I found a particular inconsistency related to this problem and opened a new issue at #36333; I hope that's ok.

Regarding your motivating issue, I do not think you will be able to get the body of `foo` to type check because the keys are generic, however the result of `foo` will have the correct behavior if you write it as so: 
```ts
console.log(c?.key && c.value.length);
```
which is abit hacky and abuses #35513.

**Comment by jtbandes** [comment]
Thanks!

Can you explain more about "I do not think you will be able to get the body of foo to type check because the keys are generic"? It seems to me that if
1. `b[i]` is known to be of type `UnionB`, and
2. `k` is known to be of type `K`, and
3. we're inside the scope of `if (b[i].key == k)`,

then `b[i]`'s type should be narrowed to `UnionB & { key: K }`.

**Comment by RyanCavanaugh** [comment]
Fixed by #36696
